File: test/orchestrator/selfbuild-gate.test.ts
import { describe, expect, it, afterAll } from 'vitest';
import { spawnSync } from 'node:child_process';
import { mkdtempSync, rmSync, mkdirSync, writeFileSync, readFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import {
  alreadyShipped,
  evaluateStarvation,
  goalSubjectItem,
  normalizeGoalText,
  readLedgerLines,
} from '../../src/orchestrator/selfbuild-gate.js';

// PRD:888 — the selfbuild driver's starvation gate (starved()) and Q27
// re-burn guard (already_shipped()) folded out of scripts/selfbuild-loop.sh
// into src/orchestrator/selfbuild-gate.ts, exposed as `devagent
// selfbuild-gate --starved / --already-shipped <goal>`. The exit codes are
// the contract (0 continue / 1 verdict / 2 unresolved, same as backlog-check
// in PRD:889), so the CLI tests spawn the real binary against ledger
// fixtures, mirroring test/backlog-check.test.ts.

const repoRoot = join(import.meta.dirname, '..', '..');
const dirs: string[] = [];
afterAll(() => {
  for (const d of dirs) rmSync(d, { recursive: true, force: true });
});

function row(loop: number, status: string, goal: string): string {
  return `{"loop":${loop},"ts":"2026-09-06T00:00:0${loop % 10}Z","status":"${status}","goal":"${goal}"}`;
}

/** Repo fixture whose .selfbuild/ledger.jsonl holds exactly `lines`. */
function ledgerRepo(lines: string[]): string {
  const repo = mkdtempSync(join(tmpdir(), 'da-gate-'));
  dirs.push(repo);
  mkdirSync(join(repo, '.selfbuild'), { recursive: true });
  writeFileSync(join(repo, '.selfbuild', 'ledger.jsonl'), lines.map((l) => l + '\n').join(''));
  return repo;
}

function gate(repo: string, ...args: string[]): { status: number | null; out: string } {
  const r = spawnSync('npx', ['tsx', join(repoRoot, 'src/cli.ts'), 'selfbuild-gate', ...args], {
    cwd: repoRoot,
    encoding: 'utf8',
    env: { PATH: process.env.PATH, HOME: process.env.HOME },
    timeout: 60_000,
  });
  return { status: r.status, out: `${r.stdout}${r.stderr}` };
}

describe('evaluateStarvation (PRD:888 port of starved())', () => {
  it('counts consecutive non-productive rows from the tail and starves at the limit', () => {
    const lines = [row(1, 'ok', 'shipped work'), ...[2, 3, 4, 5, 6].map((n) => row(n, 'failed', 'x'))];
    expect(evaluateStarvation(lines, 5)).toMatchObject({ starved: true, count: 5 });
  });

  it('a productive row breaks the streak — one short of the limit is not starved', () => {
    const lines = [row(1, 'failed', 'x'), row(2, 'ok', 'shipped'), ...[3, 4, 5, 6].map((n) => row(n, 'failed', 'x'))];
    expect(evaluateStarvation(lines, 5)).toMatchObject({ starved: false, count: 4 });
  });

  it('every documented productive class breaks: ok | pr-open | merged | pushed', () => {
    for (const status of ['ok', 'pr-open', 'merged', 'pushed']) {
      const lines = [...[1, 2, 3, 4, 5].map((n) => row(n, 'failed', 'x')), row(6, status, 'handed off')];
      expect(evaluateStarvation(lines, 5).starved).toBe(false);
    }
  });

  it('statuses merely containing a productive word do not break (push-failed counts)', () => {
    const lines = [...[1, 2, 3, 4].map((n) => row(n, 'failed', 'x')), row(5, 'push-failed', 'x')];
    expect(evaluateStarvation(lines, 5)).toMatchObject({ starved: true, count: 5 });
  });

  it('degraded rows are exempt: they neither count nor break the streak', () => {
    // 2026-09-05 class: three provider-degraded rows must not read as strikes.
    const exempt = ['operator-degraded', 'operator-diverged', 'provider-degraded'];
    const lines = [
      row(1, 'ok', 'shipped'),
      ...[2, 3, 4].map((n) => row(n, exempt[n % 3], 'pause')),
      ...[5, 6].map((n) => row(n, 'failed', 'x')),
    ];
    expect(evaluateStarvation(lines, 5)).toMatchObject({ starved: false, count: 2 });
    // A degraded tail cannot hide an already-starved streak beneath it.
    const buried = [...[1, 2, 3, 4, 5].map((n) => row(n, 'failed', 'x')), ...[6, 7].map((n) => row(n, 'provider-degraded', 'pause'))];
    expect(evaluateStarvation(buried, 5)).toMatchObject({ starved: true, count: 5 });
  });

  it('empty ledger is not starved; limit 0 starves trivially (shell -ge parity)', () => {
    expect(evaluateStarvation([], 5)).toMatchObject({ starved: false, count: 0 });
    expect(evaluateStarvation([], 0).starved).toBe(true);
  });
});

describe('alreadyShipped (PRD:888 port of already_shipped())', () => {
  const Q41_CANDIDATE =
    'Goal: Q41 — surface consecutive cross-role provider degradation. Loops 106–108 logged three `provider-degraded` rows in 75s; those rows are starvation-gate-exempt';
  // Real loop-100 row: mentions "+ Q41" only inside the parenthetical.
  const LOOP100_ROW = row(
    100,
    'ok',
    'Goal: Divergence-resilient doc-sync (PRD §17 doc-sync defect + Q41 degradation surface; loops 95–99 burned on generic divergence misclassification). In `src/git',
  );

  it('matches the real re-burn class: goal prefix on a productive row', () => {
    const goal = 'Cross-board retry memory beyond the SHA guard so the scout deprioritizes until the root-cause fix lands (Q27).';
    const lines = [row(53, 'ok', `Goal: ${goal}`)];
    expect(alreadyShipped(`Goal: ${goal}`, lines)).toEqual({ shipped: true, reason: 'goal-prefix' });
  });

  it('normalizes before matching: quotes and newlines in the candidate still hit', () => {
    const recorded = row(70, 'ok', 'Goal: fold selfbuild gates — starvation + Q27 re-burn guard into typed code with fixtures and CLI smoke coverage');
    const candidate = 'Goal: fold selfbuild gates\n\t— "starvation" + Q27 re-burn guard into typed code with fixtures and CLI smoke coverage';
    expect(normalizeGoalText(candidate)).toContain(' starvation ');
    expect(alreadyShipped(candidate, [recorded]).shipped).toBe(true);
  });

  it('subject-id match survives goal rewrites (loop-58/71 Q35 class)', () => {
    const lines = [row(110, 'ok', 'Goal: Q41 — surface consecutive cross-role provider degradation. Loops 106–108 logged three `provider-degraded` rows in 75s, and those rows are starvation-gate-')];
    expect(alreadyShipped('Goal: Q41 — surface provider degradation across roles (follow-up slice)', lines)).toEqual({
      shipped: true,
      reason: 'subject-id',
    });
  });

  it('loop-100 guard: an incidental "+ Q41" inside a parenthetical is NOT a match', () => {
    expect(alreadyShipped(Q41_CANDIDATE, [LOOP100_ROW])).toEqual({ shipped: false, reason: null });
  });

  it('non-productive rows never match (skipped/failed rows are not evidence)', () => {
    const lines = [row(109, 'skipped', Q41_CANDIDATE), row(110, 'failed', Q41_CANDIDATE)];
    expect(alreadyShipped(Q41_CANDIDATE, lines).shipped).toBe(false);
  });

  it('unrelated goal, empty goal, and missing goal marker all read as not shipped', () => {
    // Empty goal is the one intentional divergence from the awk original:
    // index($0, "") == 1 made the shell read an empty goal as shipped-on-
    // everything; the driver's ^Goal: gate makes it unreachable anyway.
    expect(alreadyShipped('Goal: something entirely new (Q99)', [LOOP100_ROW]).shipped).toBe(false);
    expect(alreadyShipped('', [LOOP100_ROW]).shipped).toBe(false);
    expect(alreadyShipped('Goal: Q41 x', []).shipped).toBe(false);
  });

  it('goalSubjectItem: id must sit before the first paren, capped at 80 chars', () => {
    expect(goalSubjectItem('Goal: Q35 cross-board retry memory (shipped as #100)')).toBe('Q35');
    expect(goalSubjectItem('Goal: doc-sync surface (PRD §17 + Q41 degradation surface)')).toBe('');
    expect(goalSubjectItem(`Goal: ${'x'.repeat(90)} Q42`)).toBe('');
    expect(goalSubjectItem('Goal: plain goal, no id')).toBe('');
  });
});

describe('selfbuild-gate CLI (exit-code contract: 0 continue / 1 verdict / 2 unresolved)', () => {
  it('--starved: exit 1 with the STARVED verdict word once the limit is reached', () => {
    const repo = ledgerRepo([1, 2, 3, 4, 5].map((n) => row(n, 'failed', 'x')));
    const r = gate(repo, '--starved', '--repo', repo);
    expect(r.status).toBe(1);
    expect(r.out).toContain('starved: 5 consecutive non-productive iterations >= limit 5');
  });

  it('--starved: exit 0 when the tail is productive; --limit overrides the default 5', () => {
    const repo = ledgerRepo([...[1, 2, 3].map((n) => row(n, 'failed', 'x')), row(4, 'pr-open', 'handed to review')]);
    expect(gate(repo, '--starved', '--repo', repo).status).toBe(0);
    const streak = ledgerRepo([1, 2, 3].map((n) => row(n, 'failed', 'x')));
    expect(gate(streak, '--starved', '--limit', '3', '--repo', streak).status).toBe(1);
    expect(gate(streak, '--starved', '--repo', streak).out).toContain('productive: 3 consecutive');
  });

  it('--already-shipped: exit 1 + "already shipped" on a productive-row match, exit 0 otherwise', () => {
    const shipped = row(110, 'ok', 'Goal: Q41 — surface consecutive cross-role provider degradation. Loops 106–108 logged three rows and those rows are starvation-gate-exempt');
    const repo = ledgerRepo([shipped]);
    const hit = gate(repo, '--already-shipped', 'Goal: Q41 — surface provider degradation across roles (follow-up slice)', '--repo', repo);
    expect(hit.status).toBe(1);
    expect(hit.out).toContain('already shipped (subject-id match on a productive ledger row)');
    const miss = gate(repo, '--already-shipped', 'Goal: Q42 brand-new work nobody recorded', '--repo', repo);
    expect(miss.status).toBe(0);
    expect(miss.out).toContain('not shipped — dispatch ok');
  });

  it('missing ledger reads as continue for both gates (shell `[ -f ] || return 1` parity)', () => {
    const repo = mkdtempSync(join(tmpdir(), 'da-gate-empty-'));
    dirs.push(repo);
    expect(gate(repo, '--starved', '--repo', repo).status).toBe(0);
    expect(gate(repo, '--already-shipped', 'Goal: anything', '--repo', repo).status).toBe(0);
  });

  it('unresolved (exit 2): no gate flag, or both flags at once', () => {
    const repo = ledgerRepo([row(1, 'failed', 'x')]);
    expect(gate(repo, '--repo', repo).status).toBe(2);
    expect(gate(repo, '--starved', '--already-shipped', 'Goal: x', '--repo', repo).status).toBe(2);
  });

  it('readLedgerLines: trailing newline yields no phantom empty row', () => {
    const repo = ledgerRepo([row(1, 'ok', 'x')]);
    const lines = readLedgerLines(join(repo, '.selfbuild', 'ledger.jsonl'));
    expect(lines).toHaveLength(1);
    expect(readLedgerLines(join(repo, 'nope.jsonl'))).toEqual([]);
  });
});

describe('selfbuild-loop.sh wiring (PRD:888 thin caller)', () => {
  const script = readFileSync(join(repoRoot, 'scripts', 'selfbuild-loop.sh'), 'utf8');

  it('dispatches the CLI gates instead of embedding awk decision logic', () => {
    expect(script).toContain(`out="$("${'${DEVAGENT[@]}'}" selfbuild-gate --starved --limit "$STARVATION_LIMIT" --repo "$REPO" 2>&1)" || rc=$?`);
    expect(script).toContain(`out="$("${'${DEVAGENT[@]}'}" selfbuild-gate --already-shipped "$1" --repo "$REPO" 2>&1)" || rc=$?`);
    expect(script).not.toContain('awk -v lim="$STARVATION_LIMIT"');
    expect(script).not.toContain('gsub(/.*goal:/');
  });

  it('honors rc 1 only with the CLI verdict word — a crashed gate must not halt/skip', () => {
    expect(script).toContain(`[ "$rc" -eq 1 ] && [[ "$out" == *"starved:"* ]]`);
    expect(script).toContain(`[ "$rc" -eq 1 ] && [[ "$out" == *"already shipped"* ]]`);
  });

  it('call sites keep the documented semantics: halt on starvation, Q27 guard after backlog-check', () => {
    expect(script).toContain('if starved; then');
    expect(script).toContain('if [ "$GUARD_RESOLVED" != 1 ] && already_shipped "$GOAL"; then');
  });
});

File: internal/orchestrator/selfbuild-gate_test.go
// Package file mirrors test/orchestrator/selfbuild-gate.test.ts (FR-GO-07, issue #194).
// The CLI-spawn (`devagent selfbuild-gate`) and shell-wiring describes are
// CliWiring-sibling scope and intentionally not ported here.
package orchestrator

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// sgRow mirrors the TS row(loop, status, goal) fixture.
func sgRow(loop int, status, goal string) string {
	return `{"loop":` + sgItoa(loop) + `,"ts":"2026-09-06T00:00:0` + sgItoa(loop%10) + `Z","status":"` + status + `","goal":"` + goal + `"}`
}

func sgItoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var b [20]byte
	i := len(b)
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		b[i] = '-'
	}
	return string(b[i:])
}

// sgLedgerRepo mirrors ledgerRepo: a repo whose .selfbuild/ledger.jsonl holds lines.
func sgLedgerRepo(t *testing.T, lines []string) (string, string) {
	t.Helper()
	repo := t.TempDir()
	dir := filepath.Join(repo, ".selfbuild")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		t.Fatal(err)
	}
	path := filepath.Join(dir, "ledger.jsonl")
	content := ""
	for _, l := range lines {
		content += l + "\n"
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		t.Fatal(err)
	}
	return repo, path
}

func TestSelfbuildGateStarvation(t *testing.T) {
	t.Run("counts consecutive non-productive rows from the tail and starves at the limit", func(t *testing.T) {
		lines := []string{sgRow(1, "ok", "shipped work"), sgRow(2, "failed", "x"), sgRow(3, "failed", "x"), sgRow(4, "failed", "x"), sgRow(5, "failed", "x"), sgRow(6, "failed", "x")}
		v := EvaluateStarvation(lines, 5)
		if !v.Starved || v.Count != 5 {
			t.Errorf("verdict = %+v, want starved/5", v)
		}
	})

	t.Run("a productive row breaks the streak — one short of the limit is not starved", func(t *testing.T) {
		lines := []string{sgRow(1, "failed", "x"), sgRow(2, "ok", "shipped"), sgRow(3, "failed", "x"), sgRow(4, "failed", "x"), sgRow(5, "failed", "x"), sgRow(6, "failed", "x")}
		v := EvaluateStarvation(lines, 5)
		if v.Starved || v.Count != 4 {
			t.Errorf("verdict = %+v, want not-starved/4", v)
		}
	})

	t.Run("every documented productive class breaks: ok | pr-open | merged | pushed", func(t *testing.T) {
		for _, status := range []string{"ok", "pr-open", "merged", "pushed"} {
			lines := []string{sgRow(1, "failed", "x"), sgRow(2, "failed", "x"), sgRow(3, "failed", "x"), sgRow(4, "failed", "x"), sgRow(5, "failed", "x"), sgRow(6, status, "handed off")}
			if EvaluateStarvation(lines, 5).Starved {
				t.Errorf("status %q must break the streak", status)
			}
		}
	})

	t.Run("statuses merely containing a productive word do not break (push-failed counts)", func(t *testing.T) {
		lines := []string{sgRow(1, "failed", "x"), sgRow(2, "failed", "x"), sgRow(3, "failed", "x"), sgRow(4, "failed", "x"), sgRow(5, "push-failed", "x")}
		v := EvaluateStarvation(lines, 5)
		if !v.Starved || v.Count != 5 {
			t.Errorf("verdict = %+v, want starved/5", v)
		}
	})

	t.Run("degraded rows are exempt: they neither count nor break the streak", func(t *testing.T) {
		// 2026-09-05 class: three provider-degraded rows must not read as strikes.
		exempt := []string{"operator-degraded", "operator-diverged", "provider-degraded"}
		lines := []string{sgRow(1, "ok", "shipped")}
		for n := 2; n <= 4; n++ {
			lines = append(lines, sgRow(n, exempt[n%3], "pause"))
		}
		lines = append(lines, sgRow(5, "failed", "x"), sgRow(6, "failed", "x"))
		v := EvaluateStarvation(lines, 5)
		if v.Starved || v.Count != 2 {
			t.Errorf("verdict = %+v, want not-starved/2", v)
		}
		// A degraded tail cannot hide an already-starved streak beneath it.
		buried := []string{sgRow(1, "failed", "x"), sgRow(2, "failed", "x"), sgRow(3, "failed", "x"), sgRow(4, "failed", "x"), sgRow(5, "failed", "x"), sgRow(6, "provider-degraded", "pause"), sgRow(7, "provider-degraded", "pause")}
		v = EvaluateStarvation(buried, 5)
		if !v.Starved || v.Count != 5 {
			t.Errorf("buried verdict = %+v, want starved/5", v)
		}
	})

	t.Run("empty ledger is not starved; limit 0 starves trivially (shell -ge parity)", func(t *testing.T) {
		if v := EvaluateStarvation(nil, 5); v.Starved || v.Count != 0 {
			t.Errorf("verdict = %+v, want not-starved/0", v)
		}
		if v := EvaluateStarvation(nil, 0); !v.Starved {
			t.Errorf("limit 0 must starve trivially, got %+v", v)
		}
	})
}

func TestSelfbuildGateShipped(t *testing.T) {
	q41Candidate := "Goal: Q41 — surface consecutive cross-role provider degradation. Loops 106–108 logged three `provider-degraded` rows in 75s; those rows are starvation-gate-exempt"
	// Real loop-100 row: mentions "+ Q41" only inside the parenthetical.
	loop100Row := sgRow(100, "ok", "Goal: Divergence-resilient doc-sync (PRD §17 doc-sync defect + Q41 degradation surface; loops 95–99 burned on generic divergence misclassification). In `src/git")

	t.Run("matches the real re-burn class: goal prefix on a productive row", func(t *testing.T) {
		goal := "Cross-board retry memory beyond the SHA guard so the scout deprioritizes until the root-cause fix lands (Q27)."
		lines := []string{sgRow(53, "ok", "Goal: "+goal)}
		v := AlreadyShipped("Goal: "+goal, lines)
		if !v.Shipped || v.Reason != ShippedReasonGoalPrefix {
			t.Errorf("verdict = %+v, want shipped/goal-prefix", v)
		}
	})

	t.Run("normalizes before matching: quotes and newlines in the candidate still hit", func(t *testing.T) {
		recorded := sgRow(70, "ok", "Goal: fold selfbuild gates — starvation + Q27 re-burn guard into typed code with fixtures and CLI smoke coverage")
		candidate := "Goal: fold selfbuild gates\n\t— \"starvation\" + Q27 re-burn guard into typed code with fixtures and CLI smoke coverage"
		if got := NormalizeGoalText(candidate); !strings.Contains(got, " starvation ") {
			t.Errorf("normalizeGoalText = %q, want to contain ' starvation '", got)
		}
		if !AlreadyShipped(candidate, []string{recorded}).Shipped {
			t.Error("normalized candidate must match")
		}
	})

	t.Run("subject-id match survives goal rewrites (loop-58/71 Q35 class)", func(t *testing.T) {
		lines := []string{sgRow(110, "ok", "Goal: Q41 — surface consecutive cross-role provider degradation. Loops 106–108 logged three `provider-degraded` rows in 75s, and those rows are starvation-gate-")}
		v := AlreadyShipped("Goal: Q41 — surface provider degradation across roles (follow-up slice)", lines)
		if !v.Shipped || v.Reason != ShippedReasonSubjectID {
			t.Errorf("verdict = %+v, want shipped/subject-id", v)
		}
	})

	t.Run("loop-100 guard: an incidental \"+ Q41\" inside a parenthetical is NOT a match", func(t *testing.T) {
		v := AlreadyShipped(q41Candidate, []string{loop100Row})
		if v.Shipped || v.Reason != ShippedReasonNone {
			t.Errorf("verdict = %+v, want not-shipped", v)
		}
	})

	t.Run("non-productive rows never match (skipped/failed rows are not evidence)", func(t *testing.T) {
		lines := []string{sgRow(109, "skipped", q41Candidate), sgRow(110, "failed", q41Candidate)}
		if AlreadyShipped(q41Candidate, lines).Shipped {
			t.Error("non-productive rows must not match")
		}
	})

	t.Run("unrelated goal, empty goal, and missing goal marker all read as not shipped", func(t *testing.T) {
		if AlreadyShipped("Goal: something entirely new (Q99)", []string{loop100Row}).Shipped {
			t.Error("unrelated goal must not match")
		}
		if AlreadyShipped("", []string{loop100Row}).Shipped {
			t.Error("empty goal must never match (intentional awk divergence)")
		}
		if AlreadyShipped("Goal: Q41 x", nil).Shipped {
			t.Error("empty ledger must read not shipped")
		}
	})

	t.Run("goalSubjectItem: id must sit before the first paren, capped at 80 chars", func(t *testing.T) {
		if got := GoalSubjectItem("Goal: Q35 cross-board retry memory (shipped as #100)"); got != "Q35" {
			t.Errorf("item = %q, want Q35", got)
		}
		if got := GoalSubjectItem("Goal: doc-sync surface (PRD §17 + Q41 degradation surface)"); got != "" {
			t.Errorf("item = %q, want empty (id in parenthetical)", got)
		}
		if got := GoalSubjectItem("Goal: " + strings.Repeat("x", 90) + " Q42"); got != "" {
			t.Errorf("item = %q, want empty (id beyond the 80-char cap)", got)
		}
		if got := GoalSubjectItem("Goal: plain goal, no id"); got != "" {
			t.Errorf("item = %q, want empty", got)
		}
	})
}

func TestSelfbuildGateExtras(t *testing.T) {
	t.Run("a dialect status does not break the streak by default, but breaks it when passed as an extra", func(t *testing.T) {
		lines := []string{sgRow(1, "failed", "x"), sgRow(2, "failed", "x"), sgRow(3, "failed", "x"), sgRow(4, "failed", "x"), sgRow(5, "judge-done", "all criteria evidenced")}
		if !EvaluateStarvation(lines, 5).Starved {
			t.Error("judge-done must count as a strike by default")
		}
		if EvaluateStarvation(lines, 5, "judge-done", "spec-refined").Starved {
			t.Error("judge-done must break the streak when passed as an extra")
		}
	})

	t.Run("extras extend the productive set without touching the degraded exemption", func(t *testing.T) {
		// provider-degraded rows still neither count nor break, extras or not —
		// the outage class the sibling awk copies miscounted as starvation.
		lines := []string{sgRow(1, "ok", "shipped")}
		for n := 2; n <= 4; n++ {
			lines = append(lines, sgRow(n, "provider-degraded", "pause"))
		}
		lines = append(lines, sgRow(5, "failed", "x"), sgRow(6, "failed", "x"))
		v := EvaluateStarvation(lines, 5, "judge-done")
		if v.Starved || v.Count != 2 {
			t.Errorf("verdict = %+v, want not-starved/2", v)
		}
	})

	t.Run("extras are regex-escaped: \"a.b\" cannot match a row whose status is \"axb\"", func(t *testing.T) {
		tail := []string{sgRow(1, "failed", "x"), sgRow(2, "failed", "x"), sgRow(3, "failed", "x"), sgRow(4, "failed", "x")}
		if !EvaluateStarvation(append(tail, sgRow(5, "axb", "x")), 5, "a.b").Starved {
			t.Error("a.b must not match axb")
		}
		if EvaluateStarvation(append(tail, sgRow(5, "a.b", "x")), 5, "a.b").Starved {
			t.Error("a.b must match a.b")
		}
	})
}

func TestSelfbuildGateReadLedgerLines(t *testing.T) {
	t.Run("readLedgerLines: trailing newline yields no phantom empty row", func(t *testing.T) {
		_, path := sgLedgerRepo(t, []string{sgRow(1, "ok", "x")})
		lines := ReadLedgerLines(path)
		if len(lines) != 1 {
			t.Errorf("lines = %d, want 1 (%q)", len(lines), lines)
		}
		if got := ReadLedgerLines(filepath.Join(filepath.Dir(path), "nope.jsonl")); len(got) != 0 {
			t.Errorf("missing ledger = %v, want empty", got)
		}
	})
}
